rclcpp/include/rclcpp/context.hpp

```diff
@@ -196,29 +204,38 @@ class Context : public std::enable_shared_from_this<Context>
    * and persist on repeated init's.
    *
    * \param[in] callback the on shutdown callback to be registered
-   * \return the callback passed, for convenience when storing a passed lambda
+   * \return the created shared pointer of callback handler
    */
   RCLCPP_PUBLIC
   virtual
-  OnShutdownCallback
-  on_shutdown(OnShutdownCallback callback);
+  OnShutdownCallbackHandle::WeakPtr
+  on_shutdown(OnShutdownCallbackHandle::OnShutdownCallbackType callback);
+
+  /// Remove a on_shutdown callback handle registered with on_shutdown.
+  /**
+   * \param[in] callback_handle the handle to be removed.
+   */
+  RCLCPP_PUBLIC
+  virtual
+  void
+  remove_on_shutdown_callback(const OnShutdownCallbackHandle::WeakPtr callback_handle);
 
-  /// Return the shutdown callbacks as const.
+  /// Return the shutdown callback handles as const.
   /**
    * Using the returned reference is not thread-safe with calls that modify
-   * the list of "on shutdown" callbacks, i.e. on_shutdown().
+   * the list of "on shutdown" callback handles, i.e. on_shutdown().
    */
   RCLCPP_PUBLIC
-  const std::vector<OnShutdownCallback> &
+  const std::unordered_set<OnShutdownCallbackHandle::SharedPtr> &
   get_on_shutdown_callbacks() const;
 
-  /// Return the shutdown callbacks.
+  /// Return the shutdown callback handles
   /**
    * Using the returned reference is not thread-safe with calls that modify
-   * the list of "on shutdown" callbacks, i.e. on_shutdown().
+   * the list of "on shutdown" callback handles, i.e. on_shutdown().
    */
   RCLCPP_PUBLIC
-  std::vector<OnShutdownCallback> &
+  std::unordered_set<OnShutdownCallbackHandle::SharedPtr> &
   get_on_shutdown_callbacks();
```

Comment:
Management of this structure is handled with `add/remove` functions, so I don't think we need read/write access to it. I'd just remove this signature. Users can make a copy with the const version of this method if they need a mutable copy of this structure.

Reply:
Okay.